tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -144,8 +144,12 @@ static void write_kernel_data(void)
 volatile int *ptr = NULL;
 #if defined(CONFIG_X86)
 volatile size_t size = MMU_PAGE_SIZE;
-#elif defined(CONFIG_ARM) && defined(CONFIG_PRIVILEGED_STACK_SIZE)
-volatile size_t size = CONFIG_ZTEST_STACKSIZE - CONFIG_PRIVILEGED_STACK_SIZE;
+#elif defined(CONFIG_ARM)
+#if defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
+volatile size_t size = POW2_CEIL(CONFIG_ZTEST_STACKSIZE);
+#else
+volatile size_t size = CONFIG_ZTEST_STACKSIZE + MPU_GUARD_ALIGN_AND_SIZE;
+#endif
 #else
```

Comment:
The actual intent of the read/write_kernel_stack tests is to try to write to the kernel/privileged stack for the user thread, not just to write below the current user stack.  The logic here assumed that the privileged stacks were carved out of the user stacks.  If we have to instead calculate their location independently, then these two tests need to be reworked accordingly.

Reply:
There is no way to calculate this now because there is no way to determine this unless we want to expose APIs to look up the privileged stack area from unprivileged threads.

Reply:
It appears that if I add -C to the gperf command line for generating the priv stacks hash table (so that it ends up in rodata), I can call _k_priv_stack_find(thread_stack) from application code and get back the priv stack pointer, and then attempting to dereference that pointer faults as expected. (That also required making thread_stack in ztest.c non-static so I could reference it from the test.) OTOH, not clear that we want to expose the priv stack locations to userspace if we can avoid it.

Reply:
FWIW, this can be seen at https://github.com/stephensmalley/zephyr/commit/622028c5e8c3ec1db61379bf67a05d657d510775 relative to your patches.  As I said though, not sure if we want to do this.

Reply:
> not clear that we want to expose the priv stack locations to userspace if we can avoid it.

I'll let you guys decide this, my feeling is that a determined hacker is gonna figure this out anyway so we should assume it's always known.

Reply:
Re-based fix for the read/write_stack tests at https://github.com/stephensmalley/zephyr/commit/2d2273cfd25afee096e5ff557e674a7b23d2df6a.  Feel free to add to this PR, either on top or squashed.

Reply:
Ok, how about this one then: https://github.com/stephensmalley/zephyr/commit/1933d4e21fff5232d43cb916e0003f6b5d70968f